src/case_encoder.cc

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.!
+
+#include "case_encoder.h"
+
+namespace sentencepiece {
+namespace normalizer {
+
+std::unique_ptr<CaseEncoder> CaseEncoder::Create(bool encodeCase, bool decodeCase, bool removeExtraWhiteSpace) {
+  if(encodeCase && decodeCase) {
+    LOG(ERROR) << "Cannot set both encodeCase=true and decodeCase=true";
+    return nullptr;
+  } else if(encodeCase) {
+    return std::unique_ptr<CaseEncoder>(new UpperCaseEncoder(removeExtraWhiteSpace));
+  } else if(decodeCase) {
+    return std::unique_ptr<CaseEncoder>(new UpperCaseDecoder());
+  } else {
+    return nullptr;
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////
+// Implements finite state automaton for regex: Uu+(sss|p|$)+Uu+(sss|p|$)+(Uu+(sss|p|$)+)+
+// std::regex seems to be causing weird issues with stack overflow etc. on Windows. 
```

Comment:
Could this be similar to PR https://github.com/marian-nmt/sentencepiece/pull/7 and be solved by defining some variables for regexes? Anyhow, FSA is probably faster, so just making you aware of Rohit's solution.

Reply:
This problem had repeated itself and we were getting system Stack Overflows, hence the need to fix it this way.